File: projects/javascript/10-http-server/solution/main.js
// 10-http-server (solution)
const readline = require('readline');

function parseRequest(raw) {
  const lines = raw.split('\r\n');
  if (!lines || lines.length === 0) return null;
  const first = lines[0].split(' ');
  return { method: first[0], path: first[1] };
}

function handleRoute(req) {
  if (req.path === '/api/health') {
    const body = JSON.stringify({ status: 'ok' });
    return `HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: ${body.length}\r\n\r\n${body}`;
  }
  if (req.path === '/hello') {
    const body = 'Hello World';
    return `HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nContent-Length: ${body.length}\r\n\r\n${body}`;
  }
  const body = 'Not Found';
  return `HTTP/1.1 404 Not Found\r\nContent-Type: text/plain\r\nContent-Length: ${body.length}\r\n\r\n${body}`;
}

async function main() {
  const rl = readline.createInterface({ input: process.stdin });
  let reqText = '';
  for await (const line of rl) {
    reqText += line + '\r\n';
  }
  const req = parseRequest(reqText);
  if (req) {
    console.log(handleRoute(req));
  }
}

main();

File: projects/rust/10-http-server/solution/main.rs
// 10-http-server (solution)
#![allow(unused_variables, dead_code, unused_mut, unused_imports)]
use std::io::{self, Read};

struct HttpRequest {
    method: String,
    path: String,
}

fn parse_request(raw: &str) -> Option<HttpRequest> {
    let mut lines = raw.split("\r\n");
    let first = lines.next()?;
    let parts: Vec<&str> = first.split_whitespace().collect();
    if parts.len() < 2 { return None; }
    Some(HttpRequest {
        method: parts[0].to_string(),
        path: parts[1].to_string(),
    })
}

fn handle_request(req: &HttpRequest) -> String {
    if req.path == "/hello" {
        let body = "Hello World";
        format!("HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nContent-Length: {}\r\n\r\n{}", body.len(), body)
    } else if req.path == "/api/status" {
        let body = "{\"status\":\"ok\"}";
        format!("HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: {}\r\n\r\n{}", body.len(), body)
    } else {
        let body = "Not Found";
        format!("HTTP/1.1 404 Not Found\r\nContent-Type: text/plain\r\nContent-Length: {}\r\n\r\n{}", body.len(), body)
    }
}

fn main() {
    let mut raw = String::new();
    if io::stdin().read_to_string(&mut raw).is_err() {
        return;
    }
    if let Some(req) = parse_request(&raw) {
        println!("{}", handle_request(&req));
    }
}
